### PR TITLE
Chi2 Reconstruction updates + dR matching + genlevel Z boson stuff

### PR DESCRIPTION
**Pull request by JanvanderLinden**
- adjusted calls of Chi2 reconstruction to match structure in CommonClassifier. With these settings only ttH will be reconstructed, ttZ is deactivated.
- added a script in `boostedAnalyzer/scripts/` to calculate dR matching efficiencies on a ntuple file
- added code to read out generator level Z boson information (currenty deactivated)
- added code to generate mass templates for Chi2 reconstructions
- added code to generate `ttH/ttbar_matched` flags for efficiency calculations of various reconstruction methods (currently deactivated)

### CODE REVIEW

File: BoostedAnalyzer/test/boostedAnalysis_ntuples-Legacy_2016_2017_2018_cfg.py [hunk 1]
```diff
@@ -73,7 +73,24 @@
             options.inputFiles=['file:///pnfs/desy.de/cms/tier2/store/user/mwassmer/TTToSemiLeptonic_TuneCP5_PSweights_13TeV-powheg-pythia8/KIT_tthbb_sl_skims_MC_v2_94X/181109_144129/0000/Skim_1.root']
         elif "2018" in options.dataEra:
             # options.inputFiles=['root://xrootd-cms.infn.it//store/mc/RunIIAutumn18MiniAOD/TTToSemiLeptonic_TuneCP5_13TeV-powheg-pythia8/MINIAODSIM/102X_upgrade2018_realistic_v15-v1/40000/EAA9C941-768D-164B-B5C1-C12306823C6E.root']
-        	options.inputFiles=['file:///pnfs/desy.de/cms/tier2/store/user/mschrode/TTToSemiLeptonic_TuneCP5_13TeV-powheg-pythia8/KIT_tthbb_skims_MC_94X_LEG_2018/190406_083857/0000/Skim_7.root']
+            #options.inputFiles=[
```
Comment (sebwieland) [inline BoostedAnalyzer/test/boostedAnalysis_ntuples-Legacy_2016_2017_2018_cfg.py:75]: Are so many default Files neccessary?

File: README.md [hunk 2]
```diff
@@ -69,7 +69,6 @@ Do for example:
       echo "WRONG CMSSW VERSION"
       return 1
     fi
-    git clone https://gitlab.cern.ch/kit-cn-cms-public/RecoLikelihoodReconstruction.git RecoLikelihoodReconstruction
```
Comment (sebwieland) [inline README.md:72]: Is this really completely deprecated now?

File: BoostedAnalyzer/src/essentialRecoVarProcessor.cpp [hunk 3]
```diff
@@ -4,7 +4,8 @@ using namespace std;
 
 essentialRecoVarProcessor::essentialRecoVarProcessor()
 {
-    pointerToRecoVars.reset(new ReconstructedVars());
+    // reconstruct ttH, reconstruct ttZ
+    pointerToRecoVars.reset(new ReconstructedVars(true, false));
```
Comment (sebwieland) [inline BoostedAnalyzer/src/essentialRecoVarProcessor.cpp:9]: maybe add Variables, to clarify what `(true, false)`  means

File: BoostedAnalyzer/src/MCMatchVarProcessor.cpp [hunk 4]
```diff
@@ -468,4 +598,107 @@ bool dfirst=true;
     vars.FillVar( "GenForwardQuark_E",forward_quark.energy());
     vars.FillVar( "GenForwardQuark_PDGID",forward_quark.pdgId());
   }
+
+
+    //if(input.genTopEvt.IsFilled()&&input.genTopEvt.TTxIsFilled()&&input.genTopEvt.IsSemiLepton()) {
```
Comment (sebwieland) [inline BoostedAnalyzer/src/MCMatchVarProcessor.cpp:603]: Whats happening here?

File: BoostedAnalyzer/src/MCMatchVarProcessor.cpp [hunk 5]
```diff
@@ -383,7 +445,75 @@ bool dfirst=true;
...
+    // get mass of dijet system of B indices were found
+    if( minDrB1<.25 && minDrB2<.25 )
+    {
+        // get dijet
+        math::XYZTLorentzVector higgs_vec = jetvecs[idxb1]+jetvecs[idxb2];
+        vars.FillVar("GenHiggsMassFromMatchedJets", higgs_vec.M());
+    }
+
   }
+
```
Comment (sebwieland) [inline BoostedAnalyzer/src/MCMatchVarProcessor.cpp:458]: Whats happening here?

File: BoostedAnalyzer/test/boostedAnalysis_ntuples-Legacy_2016_2017_2018_cfg.py (hunk 1 above)
Comment (JanvanderLinden) [reply BoostedAnalyzer/test/boostedAnalysis_ntuples-Legacy_2016_2017_2018_cfg.py:75]: Hello sebwieland,
thank you for your comment.

No, they are of course not neccessary, I think I will remove some of them.

File: README.md (hunk 2 above)
Comment (JanvanderLinden) [reply README.md:72]: Hello sebwieland,
thank you for your comment.

At the moment this module is not used anymore, as the Chi2 reconstruction was moved to the CommonClassifier itself.
If we want to activate a likelihood based reconstruction in the future, we need to re-add this module though, as this is not yet implemented in the CommonClassifier, but i would recommend to move this reconstruction to the CommonClassifier aswell.

File: BoostedAnalyzer/src/MCMatchVarProcessor.cpp (hunk 4 above)
Comment (JanvanderLinden) [reply BoostedAnalyzer/src/MCMatchVarProcessor.cpp:603]: Hello sebwieland,
thank you for your comment.

This part of the module tries to find matching reco-level jets for all objects in a ttH/ttbar sample, i.e. tries to find the b-Quarks from top decays, the light flavour quarks from the hadronic W decay and for ttH additionally both b-Quarks from the Higgs decay.
From this information the flags `Gen_ttbar/ttH_matched` are generated which can be used to generate a subsample of events where a reconstruction of ttH/ttbar is possible in general.

File: BoostedAnalyzer/src/essentialRecoVarProcessor.cpp (hunk 3 above)
Comment (JanvanderLinden) [reply BoostedAnalyzer/src/essentialRecoVarProcessor.cpp:9]: Hello sebwieland,
thank you for your comment.

I tried to clarify what these variables mean by adding the comment in the line above. 
But I can also clarify it in the function.

File: BoostedAnalyzer/src/MCMatchVarProcessor.cpp (hunk 5 above)
Comment (JanvanderLinden) [reply BoostedAnalyzer/src/MCMatchVarProcessor.cpp:458]: Hello sebwieland,
thank you for your comment.

This part of the module searches for two jets matching the generator level Higgs decay products. If two jets are found it calculates the invariant dijet mass, which can be used to generate a mass template for a mass based reconstruction method, like the implemented Chi2-Reconstruction.

**Comment by sebwieland** [comment]
Dear Jan van der Linden,
thank you very much for the nice Review of your PR.
From my side everything looks ready to be merged.
Grats,
Sebastian

**Pull request merged**